driver/src/orderbook/streamed/block_timestamp_reading.rs

```diff
@@ -22,34 +36,87 @@ impl BlockTimestampReading for Web3 {
     }
 }
 
-/// A cache for the block timestamp which avoids having to query the node in the case where we
-/// receive multiple events from the same block in a row.
+impl BlockTimestampBatchReading for Web3 {
+    fn block_timestamps(
+        &mut self,
+        block_hashes: HashSet<H256>,
+    ) -> BoxFuture<Result<Vec<(H256, u64)>>> {
+        let batched_web3 = ethcontract::web3::Web3::new(Batch::new(self.transport().clone()));
+        async move {
+            let mut result = Vec::with_capacity(block_hashes.len());
+            for chunk in Vec::from_iter(block_hashes.into_iter()).chunks(1000) {
+                let partial_result = query_block_timestamps_batched(&batched_web3, chunk).await?;
+                result.extend(partial_result);
+            }
+            Ok(result)
+        }
+        .boxed()
+    }
+}
+
+type BatchedWeb3 = ethcontract::web3::Web3<Batch<HttpTransport>>;
+async fn query_block_timestamps_batched(
+    batched_web3: &BatchedWeb3,
+    block_hashes: &[H256],
+) -> Result<Vec<(H256, u64)>> {
+    block_hashes.iter().for_each(|hash| {
+        batched_web3.eth().block(BlockId::Hash(*hash));
+    });
+    let result = batched_web3.transport().submit_batch().compat().await;
+    result
+        .with_context(|| "Batch RPC call to get block hashes failed")?
+        .into_iter()
+        .map(|response| {
+            let block: Block<H256> = serde_json::from_value(response?)?;
+            Ok((
+                block
+                    .hash
+                    .expect("blocks queried by hash should contain a hash"),
+                block.timestamp.low_u64(),
+            ))
+        })
+        .collect()
+}
+
+/// A cache for the block timestamp.
 #[derive(Debug)]
-pub struct MemoizingBlockTimestampReader<T> {
+pub struct CachedBlockTimestampReader<T> {
     inner: T,
-    hash: H256,
-    timestamp: u64,
+    cache: HashMap<H256, u64>,
 }
 
-impl<T> MemoizingBlockTimestampReader<T> {
+impl<T: BlockTimestampBatchReading + Send> CachedBlockTimestampReader<T> {
     pub fn new(inner: T) -> Self {
         Self {
             inner,
-            hash: H256::zero(),
-            timestamp: 0,
+            cache: HashMap::new(),
+        }
+    }
+
+    pub fn prepare_cache(&mut self, block_hashes: HashSet<H256>) -> BoxFuture<Result<()>> {
+        let missing_hashes = block_hashes
+            .into_iter()
+            .filter(|hash| self.cache.contains_key(&hash))
```

Comment:
should be NOT self contains, right?

Reply:
I removed *merge when green* because of this

Reply:
🤦 

<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2F2k5geABYZsReo%2Fgiphy.gif&token=great_catch" data-gitmeme-token="great_catch"><img src="https://media0.giphy.com/media/2k5geABYZsReo/giphy.gif" title="Created by gitme.me with /great_catch"/></a>